### ISSUE TITLE
Menu is closing inside tests

### ISSUE THREAD

**Issue by RichMatthews** [JedWatson/react-select#5867]
```
     <AsyncReactSelect
        defaultOptions
        loadOptions={async () => [
          { value: "Pear", label: "Pear" },
          { value: "Banana", label: "Banana" },
          { value: "Raspberry", label: "Raspberry" },
        ]}
        closeMenuOnSelect={false}
        isMulti
        onMenuClose={() => {
          console.log("called the close!!")
        }}
      />
```

In the browser is fine. I select an option, menu stays open and option populates into select

In my tests however it seems to close at this point

```
    await act(async () => {
      // assert option not selected yet
      expect(screen.queryByLabelText("Remove Dr Jimmy Carter")).toBe(null)

      // open the dropdown
      const selectDropdown = screen.getAllByRole("combobox")[0]
      await userEvent.click(selectDropdown)

      await screen.findByText("Pear")

      await userEvent.keyboard("{enter}")
      await screen.findByLabelText("Remove Pear")

      screen.debug()
    })
```

I notice that this line `await userEvent.keyboard("{enter}")` correctly selects the option but closes the menu. I verified this by seeing "called the close!!" being called in my tests. This is only called in the browser when I click outside (as expected). Anything wrong in the component or is it the testing library?

**Comment by frozentimes** [comment]
I tested this locally and as mentioned in the description I also get the same issue. 

I cloned the official repo and after spending some time digging into this I found that there is this **blurInputOnSelect** prop (whose default value depends on if device supports **TouchEvent**). In test environment this value is true. But in desktop browser this value is false (the reason why it works properly in browser). When this prop is true (in test enviroment) the blur event is triggered on the input element and hence the menu is closed and the focus is no longer with the input element so the test case fails if we try to consecutively make enter keyboard events and assert for the selected options. 

I wrote these very simple test cases to check this behavior 
Subsequently newer version of testing library - https://github.com/frozentimes/react-select-test-cases/blob/main/new.js

I wrote this test case with in the react select library docs section to test the same behavior with older testing library
https://github.com/frozentimes/react-select-test-cases/blob/main/old.js

Both test cases are passing for me.

As per the documentation the **blurInputOnSelect** is used for this

> Remove focus from the input when the user selects an option (handy for dismissing the keyboard on touch devices)

I tested it on my cell phone and as per the above documentation it works perfectly fine closing the keyboard and the menu on selecting an option. I am not sure if you can disabled the touch events though but looks like its not a problem on react-select side of things yet there is the above prop I think maybe exactly for this because of the functionality differences on touch and non touch devices possibly.

**Comment by Krishna8691** [comment]
I tested this locally and as mentioned in the description I also get the same issue.

I cloned the official repo and after spending some time digging into this I found that there is this **blurInputOnSelect** prop (whose default value depends on if device supports **TouchEvent**). In test environment this value is always true. But in desktop browser this value is false (the reason why it works properly in browser). When this prop is true (in test enviroment) the blur event is triggered on the input element after selecting an option and hence the menu is closed and the focus is no longer with the input element so the test case fails if we try to consecutively make enter keyboard events and assert for the selected options.

I wrote these very simple test cases to check this behavior
Subsequently newer version of testing library (with user-event) -https://github.com/Krishna8691/react-select-test-cases/blob/main/user-event.js

I wrote this test case with in the react select library docs section to test the same behavior with older testing library (with fire event) - https://github.com/Krishna8691/react-select-test-cases/blob/main/fire-event.js

Both test cases are passing for me (after setting the above prop to false which doesn't let the blur callback to be called).

As per the documentation the **blurInputOnSelect** is used for this
> Remove focus from the input when the user selects an option (handy for dismissing the keyboard on touch devices)

I tested it on my cell phone and as per the above documentation it works perfectly fine closing the keyboard and the menu on selecting an option. 
I am not sure if you can disabled the touch events though within the testing library but looks like its not a problem on react-select side of things yet there is the above prop which I think maybe is present exactly for this because of the functionality differences on touch and non touch devices possibly.

I don't see this mentioned anywhere in the documentation maybe worth mentioning in the documentation.

**Comment by RichMatthews** [comment]
thanks a lot for the great insight @Krishna8691 
@JedWatson are you able to help out or advise anything here? thanks

**Comment by Krishna8691** [comment]
> thanks a lot for the great insight @Krishna8691 
> @JedWatson are you able to help out or advise anything here? thanks

I don't think this is a bug or an issue as such so most likely this can be closed (tagging you @JedWatson for the same).

**Comment by Krishna8691** [comment]
I don't see any explicit details or hints about the behavior in the below linked comment of this thread in the documentation. I think and If you @JedWatson also think then let me know I would like to make a contribution towards the documentation regarding this if the behavior is just as expected.

https://github.com/JedWatson/react-select/issues/5867#issuecomment-1975123337

**Comment by trdp30** [comment]
> I tested this locally and as mentioned in the description I also get the same issue. But then I cloned the official repo and after spending some time digging into this I found that there is **blurInputOnSelect** prop (whose default value depends on if device supports **TouchEvent**). In test environment this value is always true. But in desktop browser this value is false. When this prop is true (in test enviroment) the blur event is triggered on the input element after selecting an option and hence the menu is closed and the focus is no longer with the input element so the test case fails if we try to consecutively make enter keyboard events and assert for the selected options.
> 
> To set the value of **blurInputOnSelect** react-select is using this function here -
> 
> [react-select/packages/react-select/src/utils.ts](https://github.com/JedWatson/react-select/blob/06e34882638d1526b9f5a1238bb567a3e9460ce5/packages/react-select/src/utils.ts#L324)
> 
> Line 324 in [06e3488](/JedWatson/react-select/commit/06e34882638d1526b9f5a1238bb567a3e9460ce5)
> 
>  export function isTouchCapable() { 
> and in testing environment this always returns true.
> I wrote these very simple test cases to check this behavior. Both test cases are passing for me (after setting the above prop to false which doesn't let the onBlur callback to be called). with userEvent -https://github.com/Krishna8691/react-select-test-cases/blob/main/user-event.js
> 
> with fireEvent - https://github.com/Krishna8691/react-select-test-cases/blob/main/fire-event.js
> 
> As per the documentation the **blurInputOnSelect** is used for this
> 
> > Remove focus from the input when the user selects an option (handy for dismissing the keyboard on touch devices)
> 
> I am not sure if you can disabled the touch events though within the testing library (cannot find anything about this) but looks like its not a problem on react-select side of things yet there is the above prop which I think maybe is present exactly for this because of the functionality differences on touch and non touch devices possibly not sure :)
> 
> Edit: I don't see this mentioned anywhere in the documentation maybe worth mentioning in the documentation.

Well explained @Krishna8691 , thanks for the time you have spent on it. I would love to know if there is any flag or property exposed to handle the application running env?

**Comment by trdp30** [comment]
I've found the solution for this use case and most others. Check out the [react-select-event](https://github.com/romgain/react-select-event?tab=readme-ov-file#openmenuinput-htmlelement-void) package. It has methods to handle the react-select common use cases. As well as good examples available.
If anyone needs any more help, please feel free to ping me.
